R/teal_data_module-eval_code.R

```diff
@@ -0,0 +1,124 @@
+setOldClass("teal_data_module")
+
+#' Evaluate the code in the qenv environment
+#' @name eval_code
+#' @description
+#' Given code is evaluated in the `qenv` environment of `teal_data` reactive defined in `teal_data_module`.
+#' @param object (`teal_data_module`)
+#' @inheritParams teal.code::eval_code
+#' @return Returns a `teal_data_module` object.
+#' @importMethodsFrom teal.code eval_code
+#' @importFrom methods setMethod
+NULL
+
+#' @rdname eval_code
+#' @export
+#'
+#' @examples
+#' tdm <- teal_data_module(
+#'   ui = function(id) div(id = shiny::NS(id)("div_id")),
+#'   server = function(id) {
+#'     shiny::moduleServer(id, function(input, output, session) {
+#'       shiny::reactive(teal_data(IRIS = iris))
+#'     })
+#'   }
+#' )
+#' \dontrun{
+#' eval_code(tdm, "IRIS <- subset(IRIS, Species == 'virginica')")
+#' }
+setMethod("eval_code", signature = c("teal_data_module", "character"), function(object, code) {
+  teal_data_module(
+    ui = function(id) {
+      ns <- NS(id)
+      object$ui(ns("mutate_inner"))
+    },
+    server = function(id) {
+      moduleServer(id, function(input, output, session) {
+        teal_data_rv <- object$server("mutate_inner")
+
+        if (!is.reactive(teal_data_rv)) {
+          stop("The `teal_data_module` must return a reactive expression.", call. = FALSE)
+        }
+
+        eventReactive(teal_data_rv(),
+          {
+            data <- tryCatch(teal_data_rv(), error = function(e) e)
+
+            if (inherits(data, "teal_data")) {
```

Comment:
```suggestion
            if (inherits(data, "teal_data", "qenv.error")) {
```

Reply:
This is currently handled in the next `else if`, do we really need to continue and eval when an error exists?

Reply:
The `qenv.error` mechanism is built specifically for evaluation errors such that an error rides along all subsequent evaluations and is interpreted at the end. This is why we have `eval_code` methods for `qenv.error`. I wanted to keep to that mechanism rather than handle `qenv.errors` differently. In my mind, I want extra handling for things that _aren't_ covered by this mechanism.